### PR TITLE
FE: don't lose annotations on Java primitive arrays

### PR DESCRIPTION
**Pull request by pyos**
^KT-48861 Fixed

### CODE REVIEW

**Review by petukhovv — APPROVED** [review]

File: compiler/testData/diagnostics/foreignAnnotationsTests/java8Tests/misc/annotatedPrimitiveArray.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// FIR_IDENTICAL
+// SKIP_TXT
+// FILE: J.java
+import org.jetbrains.annotations.Nullable;
+
+public interface J {
+    int @Nullable [] foo();
```
Comment (petukhovv) [inline compiler/testData/diagnostics/foreignAnnotationsTests/java8Tests/misc/annotatedPrimitiveArray.kt:7]: BTW are annotations applicable still properly with `int [] @Nullable []` or `int @Nullable [] []`?

**Comment by petukhovv** [comment]
There are 10 failed tests.

`LoadJava8TestGenerated$SourceJava$TypeUseAnnotations`:
- `testBaseClassTypeArguments`
- `testClassTypeParameterBounds`
- `testMethodTypeParameterBounds`
- `testReturnType`
- `testValueArguments`

The same in `LoadJava8TestGenerated$CompiledJava$TypeUseAnnotations`.

**Comment by pyos** [comment]
Updated the data.

File: compiler/testData/diagnostics/foreignAnnotationsTests/java8Tests/misc/annotatedPrimitiveArray.kt (hunk 1 above)
Comment (pyos) [reply compiler/testData/diagnostics/foreignAnnotationsTests/java8Tests/misc/annotatedPrimitiveArray.kt:7]: Yes, you can see that in some of the changed loadJava8 tests.

**Comment by Jianlong-Zhang-Accenture** [comment]
Good

**Comment by pyos** [comment]
@petukhovv this seems to be still relevant...

**Comment by mglukhikh** [comment]
We have finally discussed this PR today. It's simple by itself and could be acceptable. However, it's quite clear that we have a breaking change here, and as I know from you, this breaking change has at least some impact. In this case it's not allowed to add this error without warning. So we propose to implement a warning in K1, instead of an error. K2 already has an error, and it's quite fine.

**Comment by pyos** [comment]
Done - the annotations will only enhance for warnings without the `EnhanceNullabilityOfPrimitiveArrays` language feature. Not sure about the version, so I've put 1.9 in - should it be the next one instead?

File: compiler/util/src/org/jetbrains/kotlin/config/LanguageVersionSettings.kt [hunk 2]
```diff
@@ -285,6 +285,7 @@ enum class LanguageFeature(
     ProhibitAccessToEnumCompanionMembersInEnumConstructorCall(KOTLIN_1_9, kind = BUG_FIX), // KT-49110
     ReferencesToSyntheticJavaProperties(KOTLIN_1_9), // KT-8575
     CustomEqualsInInlineClasses(KOTLIN_1_9), // KT-24874
+    EnhanceNullabilityOfPrimitiveArrays(KOTLIN_1_9, kind = BUG_FIX), // KT-48861
```
Comment (mglukhikh) [inline compiler/util/src/org/jetbrains/kotlin/config/LanguageVersionSettings.kt:288]: Minor: normally we refer a LC (language committee) issue here. I created a stub for the issue: KT-54521.

File: core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/lazy/types/JavaTypeResolver.kt [hunk 3]
```diff
@@ -69,13 +70,13 @@ class JavaTypeResolver(
         val annotations = LazyJavaAnnotations(c, arrayType, areAnnotationsFreshlySupported = true)
 
         if (primitiveType != null) {
-            val jetType = c.module.builtIns.getPrimitiveArrayKotlinType(primitiveType)
-
-            jetType.replaceAnnotations(Annotations.create(annotations + jetType.annotations))
+            val kotlinType = c.module.builtIns.getPrimitiveArrayKotlinType(primitiveType).let {
```
Comment (mglukhikh) [inline core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/lazy/types/JavaTypeResolver.kt:73]: Are changes in this file just a code cleanup, or you expect some change of semantics here?

**Comment by mglukhikh** [comment]
> Done - the annotations will only enhance for warnings without the `EnhanceNullabilityOfPrimitiveArrays` language feature. Not sure about the version, so I've put 1.9 in - should it be the next one instead?

Yes, I'd say it should be the next version. However, it's not yet created, so I suggest to leave version 1.9 in your code as is. I will change the language feature version before merging.

File: core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/lazy/types/JavaTypeResolver.kt (hunk 3 above)
Comment (pyos) [reply core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/lazy/types/JavaTypeResolver.kt:73]: This is the main part of the fix - `replaceAnnotations` returns a new type, it does not modify the existing one.

File: compiler/util/src/org/jetbrains/kotlin/config/LanguageVersionSettings.kt (hunk 2 above)
Comment (pyos) [reply compiler/util/src/org/jetbrains/kotlin/config/LanguageVersionSettings.kt]: done

File: compiler/testData/diagnostics/foreignAnnotationsTests/java8Tests/misc/annotatedPrimitiveArray.kt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// SKIP_TXT
```
[the remark this answers is not in the document]
Comment (pyos) [reply compiler/testData/diagnostics/foreignAnnotationsTests/java8Tests/misc/annotatedPrimitiveArray.kt:2]: done

**Review by mglukhikh — APPROVED** [review]

**Comment by mglukhikh** [comment]
Note: I performed remote run, all test passed.

**Comment by mglukhikh** [comment]
@dzharkov approved this PR, so I will merge it this week. 

@pyos Could you please fill the LC issue in KT-54521? I was told you know how to do it.

**Comment by mglukhikh** [comment]
Finally merged. A lot of thanks :+1:

**Pull request closed without merging**